evm/src/cpu/dup_swap.rs

```diff
@@ -8,22 +8,25 @@ use plonky2::plonk::circuit_builder::CircuitBuilder;
 
 use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
 use crate::cpu::columns::{CpuColumnsView, MemoryChannelView};
-use crate::cpu::membus::NUM_GP_CHANNELS;
 use crate::memory::segments::Segment;
 
 /// Constrain two channels to have equal values.
+/// Since it may involve the next row and can be called at the last row
+/// (if the last instruction is a DUP), we make it a transition constraint.
 fn channels_equal_packed<P: PackedField>(
     filter: P,
     ch_a: &MemoryChannelView<P>,
     ch_b: &MemoryChannelView<P>,
     yield_constr: &mut ConstraintConsumer<P>,
 ) {
     for (limb_a, limb_b) in izip!(ch_a.value, ch_b.value) {
-        yield_constr.constraint(filter * (limb_a - limb_b));
+        yield_constr.constraint_transition(filter * (limb_a - limb_b));
     }
 }
```

Comment:
I would perhaps clarify in a comment why it's okay to leave the memory channels unconstrained if `DUP` is the last instruction.

Reply:
Or I think we can just turn it back into a normal constraint since [the padding PR](https://github.com/0xPolygonZero/plonky2/pull/1234) has been merged.